### ISSUE TITLE
Run time error with backward()  after upgrading to pytorch 1.2.0

### ISSUE THREAD

**Issue by kikiloverafa** [pytorch/pytorch#24853]
## 🐛 Bug

The following error occurs after I upgrade to Pytorch 1.2.0. 
one of the variables needed for gradient computation has been modified by an inplace operation: [torch.FloatTensor [3]], which is output 0 of AddBackward0, is at version 1; expected version 0 instead.

## To Reproduce

Here are my codes. They work fine in Pytorch 1.1.0.

import torch

b0 = torch.tensor(5.0, requires_grad = True)
a = torch.Tensor([1,0,0])
b = torch.zeros(3, requires_grad = True)
c = torch.ones(3, requires_grad = True)

bb = b+1

bb[a!=0] = torch.min(bb[a!=0], c[a!=0])

d = torch.sum(bb)
d.backward()

## Expected behavior

It should run with no error.

## Environment

torch.autograd.set_detect_anomaly(True) shows that error is with the following line.

bb[a!=0] = torch.min(bb[a!=0], c[a!=0])

 - PyTorch Version: 1.2.0
 - OS (e.g., Linux): Windows 10
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.7.3
 - CUDA/cuDNN version: None

cc @ezyang @gchanan @zou3519 @SsnL @albanD

**Comment by gchanan** [comment]
I'm guessing this is due to one of @yf225's changes, but this looks legitimate?  At least, it looks legitimate given how autograd works.

**Comment by zou3519** [comment]
I think the problem is that PyTorch 1.2 is saving `bb` for backward, while PyTorch 1.1 didn't need to. The culprit is this line in derivatives.yaml:

https://github.com/pytorch/pytorch/blob/6100205eb8a88090581a905e595e7d1d2a8c232d/tools/autograd/derivatives.yaml#L389-L391

the `index` function saves `self` for backward because it is present here. However, we don't need to save the entire tensor; in PyTorch 1.1, we only saved the geometry of `self`. (zeros_like(self) in the below snippet is special-cased in our code gen to only save the geometry of the tensor).
https://github.com/pytorch/pytorch/blob/0b868b19063645afed59d6d49aff1e43d1665b88/tools/autograd/derivatives.yaml#L383-L385

IndexBackward shouldn't need to rely on self's data, so changing it to take `zeros_like(self)` should fix the problem. I'll test a fix soon.

**Comment by kikiloverafa** [comment]
I am going to switch back to Pytorch 1.1.0 for now. Would you please let us know when this issue is fixed?

**Comment by baobablyh** [comment]
Looks like I hit the same issue. Before we get to the fix, is there any quick workaround helping unblock? Thanks.

**Comment by baobablyh** [comment]
Confirmed that building from source with zeros_like solved the issue. It worked for me.

**Comment by kikiloverafa** [comment]
> Confirmed that building from source with zeros_like solved the issue. It worked for me.

FYI: The issue no longer exists after upgrading to Pytorch 1.3.0.
